### PR TITLE
Alertmanager: Add state size limit

### PR DESCRIPTION
**Pull request by titolins**
Similar to #9402 , but for Grafana state

Fixes https://github.com/grafana/alerting-squad/issues/933

### CODE REVIEW

**Review by santihernandezc — COMMENTED** [review]
Should we add an entry to the changelog as you did in [your last PR](https://github.com/grafana/mimir/pull/9402/files)?

**Review by santihernandezc — APPROVED** [review]
LGTM 👍

**Review by tacole02 — APPROVED** [review]

**Comment by santihernandezc** [comment]
@titolins is OoO, I'm fixing some issues (mostly conflicts with `main` in `CHANGELOG.md`) before merging this PR.

File: CHANGELOG.md [hunk 1]
```diff
@@ -25,7 +25,9 @@
   * `-query-scheduler.grpc-client-config.grpc-compression=s2`
   * `-ruler.client.grpc-compression=s2`
   * `-ruler.query-frontend.grpc-client-config.grpc-compression=s2`
-* [FEATURE] Alertmanager: limit added for maximum size of the Grafana configuration (`-alertmanager.max-config-size-bytes`). #9402
+<<<<<<< HEAD
```
Comment (bboreham) [inline CHANGELOG.md:28]: Spurious

File: pkg/alertmanager/api.go [hunk 2]
```diff
@@ -90,7 +91,6 @@ func (am *MultitenantAlertmanager) GetUserConfig(w http.ResponseWriter, r *http.
 		TemplateFiles:      alertspb.ParseTemplates(cfg),
 		AlertmanagerConfig: cfg.RawConfig,
 	})
-
```
Comment (bboreham) [inline pkg/alertmanager/api.go:93]: ?

**Review by bboreham — COMMENTED** [review]
Couple of spurious changes seem to have crept in.

**Comment by titolins** [comment]
squashing the commits to make rebasing easier since CHANGELOG changes are relatively common

File: docs/sources/mimir/configure/configuration-parameters/index.md [hunk 3]
```diff
@@ -3651,8 +3651,8 @@ The `limits` block configures default and per-tenant limits imposed by component
 # CLI flag: -alertmanager.notification-rate-limit-per-integration
 [alertmanager_notification_rate_limit_per_integration: <map of string to float64> | default = {}]
 
-# Maximum size of the Grafana configuration file for Alertmanager that a tenant
-# can upload via the Alertmanager API. 0 = no limit.
+# Maximum size of the Grafana Alertmanager configuration for a tenant. 0 = no
+# limit.
```
Comment (santihernandezc) [inline docs/sources/mimir/configure/configuration-parameters/index.md:3655]: Nit: 
```suggestion
# Maximum size of the Grafana Alertmanager configuration for a tenant. 0 = no limit.
```

File: docs/sources/mimir/configure/configuration-parameters/index.md [hunk 4]
```diff
@@ -3661,6 +3661,10 @@ The `limits` block configures default and per-tenant limits imposed by component
 # CLI flag: -alertmanager.max-config-size-bytes
```
Comment (santihernandezc) [inline docs/sources/mimir/configure/configuration-parameters/index.md:3661]: Just two lines above this one, the old description for the config size limit should be fixed.

File: docs/sources/mimir/manage/mimir-runbooks/_index.md [hunk 5]
```diff
@@ -2386,6 +2386,16 @@ How to **fix** it:
 
 This error only occurs when an administrator has explicitly define a blocked list for a given tenant. After assessing whether or not the reason for blocking one or multiple queries you can update the tenant's limits and remove the pattern.
 
+### err-mimir-alertmanager-max-grafana-config-size
+
+This non-critical error occurs when the Alertmanager receives a set configuration request that contains a configuration which exceeds the configured size limit.
+The limit protects the system’s stability from potential abuse or mistakes. To configure the limit on a per-tenant basis, use the `alertmanager.max-grafana-config-size-bytes` option.
```
Comment (santihernandezc) [inline docs/sources/mimir/manage/mimir-runbooks/_index.md:2392]: Suggestion:
```suggestion
This non-critical error occurs when the Alertmanager receives a Grafana Alertmanager configuration larger than the configured size limit.
The limit protects the system’s stability from potential abuse or mistakes. To configure the limit on a per-tenant basis, use the `alertmanager.max-grafana-config-size-bytes` option.
```

File: docs/sources/mimir/manage/mimir-runbooks/_index.md [hunk 6]
```diff
@@ -2386,6 +2386,16 @@ How to **fix** it:
...
+
+### err-mimir-alertmanager-max-grafana-state-size
+
+This non-critical error occurs when the Alertmanager receives a set state request that contains a state which exceeds the configured size limit.
+The limit protects the system’s stability from potential abuse or mistakes. To configure the limit on a per-tenant basis, use the `alertmanager.max-grafana-state-size-bytes` option.
```
Comment (santihernandezc) [inline docs/sources/mimir/manage/mimir-runbooks/_index.md:2397]:
```suggestion
This non-critical error occurs when the Alertmanager receives Grafana Alertmanager state larger than the configured size limit.
The limit protects the system’s stability from potential abuse or mistakes. To configure the limit on a per-tenant basis, use the `alertmanager.max-grafana-state-size-bytes` option.
```

**Review by santihernandezc — APPROVED** [review]
LGTM! Please take a look at my comments.

File: docs/sources/mimir/configure/configuration-parameters/index.md (hunk 4 above)
Comment (titolins) [reply docs/sources/mimir/configure/configuration-parameters/index.md:3660]: This is the description of the remote AM configuration limit. I had copied the grafana config limit from this, that's why it's almost the same 😅 
I've adjust it now to keep the same wording for both

File: docs/sources/mimir/configure/configuration-parameters/index.md (hunk 3 above)
Comment (titolins) [reply docs/sources/mimir/configure/configuration-parameters/index.md:3655]: This file is auto-generate (`make doc`)

**Review by 56quarters — APPROVED** [review]
Thanks for making that change!

**Pull request merged**